### DISCUSSION TITLE
Denial of service resistance

### DISCUSSION THREAD

**Discussion by justinmoon** [fedimint/fedimint#1440] [General]
We've had some discussions on the dev calls recently about denial of service attacks. Here is an issue to track these discussions.

Relevant comment from @jkitman recently: https://github.com/fedimint/fedimint/issues/801#issuecomment-1282504209

Related issues: 
- https://github.com/fedimint/fedimint/issues/229
- https://github.com/fedimint/fedimint/issues/162

**Comment by jkitman** [comment]
Since we have a blinded mint there are some interesting solutions to Sybil attacks. Obviously the most simple is for the federation or gateway to charge ecash fees, but that is a problem if the user doesn't yet have any bitcoin or ecash and needs to receive over LN.

We could issue blinded "reputation tokens" for users that can prove they are not a Sybil attacker when receiving over LN for the first time:
- Vetted by a guardian or other user (i.e. shareable QR codes to join)
- Owning a certain amount of bitcoin UTXOs
- Has a unique Android or IOS root certificate
- Downloaded a unique copy of the client app
- Requests from a unique IP address

Hashcash probably won't work, especially for mobile devices. I can trivially generate millions of equivalent hashcash tokens spending $100 in AWS compared to a mobile phone. Proof-of-bandwidth or proof-of-storage might fare slightly better. We could do that by requiring the user to download a large number of unique nonces from the federation then require them to prove they are storing a number of them.

^ for reference copied from https://github.com/fedimint/fedimint/issues/801#issuecomment-1282504209

**Comment by justinmoon** [comment]
The federation should not allow more than 9 tokens of a given denominations. So a user can't just attempt to verify 10k 1msat tokens. Not sure if this is the case currently.

**Comment by justinmoon** [comment]
@waxwing mentioned https://privacypass.github.io/ in his [recent ecash talk](https://www.youtube.com/watch?v=VwMzNE1D3so)

**Comment by jkitman** [comment]
> The federation should not allow more than 9 tokens of a given denominations. So a user can't just attempt to verify 10k 1msat tokens. Not sure if this is the case currently.

Currently we don't check this server-side, although we may want to allow more than 9 tokens per denomination for privacy reasons.  DoS attacks could always spam multiple txs anyway.  Better to charge fees or reputation tokens when spam is high.

**Comment by NicolaLS** [comment]
Consolidation of #801:
- incoming (receive) is not a attack vector for the gw, because it can claim a refund at any time
- outgoing (send) is an attack vector for the gw but as @justinmoon suggested not more relevant than in the LN in general. We thought about solving this with reputation tokens, another approach would be hold fees.

**Comment by elsirion** [comment]
As pointed out by @justinmoon on a call recently our highest priority should be protecting against DoS by malicious users, malicious guardians are a separate topic. There also isn't a perfect anti-DoS solution and we'll have to adapt. These two points lead me to the conclusion that we should build DoS protection on the API level of each guardian and ideally not bake it into `fedimintd` since it's not consensus relevant.

To achieve that goal guardians could issue single-sig e-cash that's far more efficient to generate and validate. The architecture I'm contemplating would consist of three services:
* **fedimintd**: providing the current, not rate-limited API but only listening internally
* **Proxy**: listening on a public interface, receiving requests, checking if they contain anti-DoS-e-cash and forwarding them to `fedimintd`
* **Anti-DoS token issuer**: a service from which users can acquire the anti-DoS-e-cash. These tokens could be issued in return for:
  * Proof of burn inside the federation (basically the same as a payment to all guardians, but removes the need for an e-cash wallet in the service)
  * Proof of work: should be rather expensive, but is necessary for bootstrapping public federations
  * Proof of authority: for private federations guardians could whitelist users that can claim a certain number of these tokens per time frame

This architecture allows us to remove/add ways of acquiring these anti-DoS-tokens without touching `fedimintd`. We could consider baking the proxy into `fedimintd` to avoid deployment complexity.

EDIT: I think this architecture isn't novel, I remember this being discussed for Tor a while ago, maybe @oleonardolima knows what they ended up with.

EDIT 2: For the light-weight, single-sig e-cash we could likely use Ruben Somsen's scheme also used in Cashu https://github.com/cashubtc/nuts/blob/main/00.md#protocol

**Reply by elsirion** [reply]
Ideas from call with @dpc:
* Make verification not a proxy, but an external verification service?
  * Easier to make fault tolerant
